Stand alone

gpolumbo-broad opened the pull request:
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

dvoet committed:
stand alone docker compose for sam

gpolumbo-broad committed:
SAM Stand-alone config changes.  Things mostly start, but the sam.conf isn't right yet

gpolumbo-broad committed:
SAM Stand-alone config changes.  Things mostly start, but the sam.conf isn't right yet

gpolumbo-broad committed:
SAM Stand-alone instructions in README for SSL Certificate

gpolumbo-broad committed:
Removed external network configuration from docker-compose

gpolumbo-broad committed:
Removed reference to network config in README

gpolumbo-broad commented on docker/stand-alone/docker-compose.yaml:
```diff
@@ -0,0 +1,34 @@
+version: '2'
+services:
+  sam-app:
+    image: broadinstitute/sam:stand-alone
```
Before merging, we will need to update this to use the correct image.  (Is sam:dev the right image?)

gpolumbo-broad committed:
Cleaned up emailDomain config field in sam.conf

coverbeck commented on docker/stand-alone/README.md:
```diff
@@ -0,0 +1,68 @@
+### Email Domain
+
+Specify the email domain for your SAM instance in `docker/stand-alone/sam.conf` by changing the 
+`emailDomain = "replace.me.with.your.domain"` field.
```
Shouldn't this step be before "Starting the application"?

Also, it's not clear to me what this value should be? Should it be `sam.example.localhost.org`? If so, could you not default to that? If not, can you provide an example of what it should be?

coverbeck commented on docker/stand-alone/README.md:
````diff
@@ -0,0 +1,68 @@
+
+Add the following to your `/etc/hosts` file:
+
+```127.0.0.1 sam.example.localhost.org```
````
Most people will probably already have `127.0.0.1` in their `/etc/hosts` file, or maybe I'm assuming based on my own hosts file. So I suggest an example with that as well or instead, because you probably don't want to kill your localhost mapping.

```
127.0.0.1 localhost sam.example.localhost.org
```

coverbeck commented on docker/stand-alone/docker-compose.yaml:
```diff
@@ -0,0 +1,34 @@
+version: '2'
+services:
+  sam-app:
+    image: broadinstitute/sam:stand-alone
```
Not crucial, although it would be helpful, and it may not even belong here, to give instructions on how to build and run SAM from local sources. I see in the main README.md that there is the `./docker/build.sh jar -d build`, but tying it all together explicitly on how I build SAM locally then run that local Docker image would be helpful.

I believe I will need to do my own PR to add a new resource type to SAM, so I'm anticipating that. :)

gpolumbo-broad replied:
Will update with these instructions

coverbeck commented:
After following the instructions, I got the SAM Swagger UI to come up, but saw two issues, both done via Swagger UI html page:

1. Invoking `/status` returns a 500, with the following response body:

```
{
  "ok": false,
  "systems": {
    "OpenDJ": {
      "ok": false,
      "messages": [
        "could not find group All_Users in opendj"
      ]
    }
  }
}
```

2. Trying to list resource types, `/api/resourceTypes`, returns a 401 with the following response body:

```<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>401 Unauthorized</title>
</head><body>
<h1>Unauthorized</h1>
<p>This server could not verify that you
are authorized to access the document
requested.  Either you supplied the wrong
credentials (e.g., bad password), or your
browser doesn't understand how to supply
the credentials required.</p>
<hr>
<address>Apache Server at sam.example.localhost.org Port 29443</address>
</body></html>
```

I was able to authorize in Swagger before doing that, the red exclamation point had changed to the blue `i`.

MatthewBemis commented:
@coverbeck you can fix the 401 by hitting the POST /register/user endpoint to get your account registered in sam. we should add a note about this at the end of the setup instructions.

coverbeck commented:
@MatthewBemis Thanks for the tip. Just tried the POST and a subsequent GET `/register/user` returns a 200 with my email in the response body, so it worked.

However, even after that, I continue to get a 401 on GET `/api/resourceTypes`.

Charles

gpolumbo-broad committed:
GAWB-3606 Updated README, extracted customization variables out to a .env file, modified application to work at https://localhost by default

gpolumbo-broad commented:
@coverbeck we're seeing the same 401 and are working to get it sorted.

gpolumbo-broad committed:
GAWB-3606 Added instructions for registering your user account.  Disabled the audience verification in site.conf that users should reenable if running in a 'real' environment

gpolumbo-broad committed:
GAWB-3606 Updated docker/stand-alone/README to include instructions for building and running SAM with customized code

MatthewBemis reviewed (approved):
👍

gpolumbo-broad commented:
@coverbeck I believe I have addressed your comments, of course please let us know if there's anything else that doesn't work or could be simplified.

- App now defaults to run on localhost
- All config file settings are now isolated to .env file
- Set more default values so that there's less that you MUST specify to get SAM running
- Instructions added for building custom SAM app and docker image
- Fixed the 401 errors

coverbeck commented:
Thanks @gpolumbo-broad !

I'll give it a shot and let you know.

coverbeck commented on docker/stand-alone/.env:
```diff
@@ -0,0 +1,13 @@
+
+# These values MUST be changed in order to be able to authenticate with SAM
+GOOGLE_OAUTH_CLIENT_ID=change_me
+GOOGLE_CLOUD_PROJECT_DOMAIN=change_me
```
What is the `GOOGLE_CLOUD_PROJECT_DOMAIN`? Is it the project id in the GCP console? It doesn't seem like it, but I can't tell what it should be.

coverbeck replied:
I tried setting it to my the domain of my organization, ucsc.edu, and it seems to be working.

coverbeck commented on docker/stand-alone/.env:
```diff
@@ -0,0 +1,13 @@
+# Variables below can optionally be modified as needed
+SAM_APP_IMAGE=broadinstitute/sam:stand-alone
+SAM_HOST=localhost
+EMAIL_DOMAIN=example.foo
```
What is this used for and what should the value be? Looks like it's optional so it may not matter, but it would be nice to understand.

dvoet replied:
we use email addresses as alternate identifiers, in production there is a gsuite domain to make them real email addresses. Here it doesn't matter.

coverbeck commented on docker/stand-alone/README.md:
```diff
@@ -0,0 +1,84 @@
+1. Click the `Authorize` button in the upper right of the page to log into SAM
+    1. In the window that pops up, check the 3 checkboxes (`openid`, `email`, and `profile`)
+    1. Click the `Authorize` button at the bottom of the pop-up window
+1. Register your user account in the **Users** section by expanding the `POST /register/user` operation and clicking the
```
When I did this, I got an error saying the user had already been registered, because I had run this previously against a previous version of the PR.

That said, although maybe this is getting too much into documenting Docker, maybe it would be worth mentioning how to clear the volumes, especially if a user wanted to start from scratch.

gpolumbo-broad replied:
The easiest way to start over is to run `docker-compose down; docker-compose up`

coverbeck replied:
I think you need to do `docker-compose down -v` if you don't want users persisted between sessions. But maybe this is getting too much into details and most Docker users will know this already (I just learned about it fairly recently).

gpolumbo-broad committed:
GAWB-3569 Changed name of GOOGLE_CLOUD_PROJECT_ID

gpolumbo-broad commented:
jenkins retest

coverbeck commented on docker/stand-alone/README.md:
```diff
@@ -0,0 +1,84 @@
+1. Run `sbt assembly` - This should build a `.jar` in `target/scala-2.12`
+1. Copy the `.jar` to SAM's root directory: `cp target/scala-2.12/sam-assembly-0.1-xxxxxxx-SNAPSHOT.jar .`
+1. [Re-build](https://docs.docker.com/engine/reference/commandline/build/) the `sam-app` Docker container by running 
+`docker build -t sam/my-tag`.  You can name the tag whatever you want by changing the argument passed to `-t`. 
```
Should be `docker build -t sam/mytag .` (need a period inside the backtick).

gpolumbo-broad commented:
Once again, Jenkins retest please

gpolumbo-broad committed:
GAWB-3606 Fixed placement of period in README

gpolumbo-broad commented:
reported unit test failure:  https://fc-jenkins.dsp-techops.broadinstitute.org/job/sam-PR-fiab-test/272/

jenkins retest
